arch/x86/core/intel64/userspace.S

```diff
@@ -21,17 +68,39 @@
 z_x86_syscall_entry_stub:
 	swapgs
 
-	/* Switch to the privilege mode stack pointer stored in
-	 * x86_tss64.psp and store the user mode stack pointer in
-	 * x86_tss64.usp, immediately pushing it once the stack switch
-	 * is done since this is a per-cpu and not per-thread area.
-	 *
-	 * This dance is necessary as upon entry we have no free registers
-	 * nor a stack we can push to.
+	/* Save original stack pointer from user mode in memory, at the
+	 * moment we have no free registers or stack to save it to. This
+	 * eventually gets put on the stack before we re-enable interrupts
+	 * as this is a per-cpu and not per-thread area.
 	 */
 	movq	%rsp, %gs:__x86_tss64_t_usp_OFFSET
```

Comment:
question, can't we use `x86_tss64.rsp2` for this purpose and get rid of `usp` pointer ?

Reply:
the CPU will not use it, stack switching for syscall/sysret is done manually in software.

Reply:
also rsp2 is used by the CPU for privilege elevations to Ring 2 if IST is not configured in the IDT for that vector. we don't use ring 2, and we have IST set.

Reply:
> the CPU will not use it, stack switching for syscall/sysret is done manually in software.

I meant to use it as an scratch pointer, if I understood correctly you are using `tss.usp` to save rsp address and then restore it. Couldn't we remove this pointer in tss struct and just use rsp2 to the same purpose since we don't use ring2 and the hw should not use it ?

Btw, I don't know about IST relationship so I may be completely wrong :) Will read about it.

Reply:
yeah that could be done, but I thought this could potentially confuse people...they would see a value in rsp2 and wonder what we're doing elevating to ring 2

we're only talking about a few bytes here on a platform which has a lot of RAM relatively speaking

Reply:
yeah, don't worry about it, it is probably better this way :) It was more an exercise to me to understand it !